File: WordleArena/Api/Hubs/RoomApi.cs
using WordleArena.Api.Hubs.Messages;
using WordleArena.Api.Hubs.Messages.ClientToServer;
using WordleArena.Domain;
using WordleArena.Domain.Events.UserSession;
using WordleArena.Domain.Grains;

namespace WordleArena.Api.Hubs;

public partial class GameHub
{
    public async Task<ApiResponse<object?>> UpdateRoomPlayerReadiness(RoomPlayerUpdateReadyRequest request)
    {
        var userId = HubUtils.AssertUserVerifiedAndGetPlayerId(Context);
        var roomGrain = factory.GetGrain<IRoomGrain>(request.RoomId.Id);
        var status = await roomGrain.GetRoomStatus();
        if (status != RoomStatus.InProgress) return new ApiResponse<object?>(ApiResponseStatusCode.NotFound, null);

        var (success, message) = await roomGrain.UpdatePlayerReadiness(userId, request.IsReady);
        if (success) return new ApiResponse<object?>(ApiResponseStatusCode.Ok, null, message);
        return new ApiResponse<object?>(ApiResponseStatusCode.Forbidden, null, message);
    }

    public async Task<ApiResponse<RoomId?>> CreateARoom()
    {
        var userId = HubUtils.AssertUserVerifiedAndGetPlayerId(Context);
        var roomId = RoomId.GenerateNewRoomId();
        var roomGrain = factory.GetGrain<IRoomGrain>(roomId.Id);
        await roomGrain.Start(roomId, GameType.Tempo);
        var (success, message) = await roomGrain.Join(userId);
        if (success)
            return new ApiResponse<RoomId?>(ApiResponseStatusCode.Ok, roomId, message);
        return new ApiResponse<RoomId?>(ApiResponseStatusCode.Forbidden, null, message);
    }

    public async Task<ApiResponse<object?>> LeaveRoom(RoomId roomId)
    {
        var userId = HubUtils.AssertUserVerifiedAndGetPlayerId(Context);
        var roomGrain = factory.GetGrain<IRoomGrain>(roomId.Id);
        var status = await roomGrain.GetRoomStatus();
        if (status != RoomStatus.InProgress) return new ApiResponse<object?>(ApiResponseStatusCode.NotFound, null);
        var (success, message) = await roomGrain.Leave(userId);
        if (success)
        {
            await mediator.Publish(new UserLeftRoom(userId, roomId));

            return new ApiResponse<object?>(ApiResponseStatusCode.Ok, null, message);
        }

        return new ApiResponse<object?>(ApiResponseStatusCode.Forbidden, null, message);
    }

    public async Task<ApiResponse<RoomState?>> GetRoomState(RoomId roomId)
    {
        var userId = HubUtils.AssertUserVerifiedAndGetPlayerId(Context);
        var roomGrain = factory.GetGrain<IRoomGrain>(roomId.Id);
        var status = await roomGrain.GetRoomStatus();
        if (status != RoomStatus.InProgress) return new ApiResponse<RoomState?>(ApiResponseStatusCode.NotFound, null);

        var isParticipant = await roomGrain.IsPlayerInRoom(userId);
        if (isParticipant)
        {
            var state = await roomGrain.GetRoomState();
            return new ApiResponse<RoomState?>(ApiResponseStatusCode.Ok, state);
        }

        return new ApiResponse<RoomState?>(ApiResponseStatusCode.Forbidden, null,
            "You are not a participant of a room");
    }

    public async Task<ApiResponse<object?>> JoinRoom(JoinRoomRequest request)
    {
        var userId = HubUtils.AssertUserVerifiedAndGetPlayerId(Context);
        var roomGrain = factory.GetGrain<IRoomGrain>(request.RoomId.Id);
        var status = await roomGrain.GetRoomStatus();
        if (status != RoomStatus.InProgress) return new ApiResponse<object>(ApiResponseStatusCode.NotFound, null);
        var (success, message) = await roomGrain.Join(userId);

        if (success)
        {
            await mediator.Publish(new UserJoinedRoom(userId, request.RoomId));
            return new ApiResponse<object>(ApiResponseStatusCode.Ok, null, message);
        }

        return new ApiResponse<object?>(ApiResponseStatusCode.Forbidden, null, message);
    }
}

File: WebApp/src/lib/clients/room-api.ts
import type {ApiResponse} from "$lib/generate/messages/api-response.ts";
import type {RoomId} from "$lib/generate/domain/room-id.ts";
import {getGameClient} from "$lib/clients/gameclient.ts";
import type {RoomPlayerUpdateReadyRequest} from "$lib/generate/messages/room-player-update-ready-request.ts";
import type {RoomState} from "$lib/generate/domain/room-state.ts";
import type {JoinRoomRequest} from "$lib/generate/messages/join-room-request.ts";


export async function CreateARoom(): Promise<ApiResponse<RoomId | null>> {
    let client = await getGameClient()
    return await client.getConnection().invoke<ApiResponse<RoomId | null>>("CreateARoom");
}

export async function updateRoomPlayerReadiness(request: RoomPlayerUpdateReadyRequest): Promise<ApiResponse<null>> {
    let client = await getGameClient();
    return await client.getConnection().invoke<ApiResponse<null>>("UpdateRoomPlayerReadiness", request);
}

// Leave an existing room
export async function leaveRoom(roomId: RoomId): Promise<ApiResponse<null>> {
    let client = await getGameClient();
    return await client.getConnection().invoke<ApiResponse<null>>("LeaveRoom", roomId);
}

// Get the current state of a room
export async function getRoomState(roomId: RoomId): Promise<ApiResponse<RoomState | null>> {
    let client = await getGameClient();
    return await client.getConnection().invoke<ApiResponse<RoomState | null>>("GetRoomState", roomId);
}

// Join an existing room
export async function joinRoom(request: JoinRoomRequest): Promise<ApiResponse<null>> {
    let client = await getGameClient();
    return await client.getConnection().invoke<ApiResponse<null>>("JoinRoom", request);
}